src/main/kotlin/exercise4/task2/TransactionalBankAccount.kt

```diff
@@ -0,0 +1,225 @@
+    account.deposit(1000.0)
+    account.withdraw(500.0)
+    account.deposit(-100.0)
+    account.withdraw(2000.0)
+    account.deposit(500.0)
+    account.withdraw(100.0)
+
+    // Displaying updated account information
+    account.displayAccountInfo()
+
+    println("All Transactions Dates:")
+    account.getAllTransactions().forEach { println(it.getTransactionDate().prettyPrint()) }
+
+    println("\nTransactions with Amount > 600:")
+    account.getAllTransactionsBy { it.getAmount() > 600 }.forEach{ println(it.getAmount()) }
+
+    println("\nTransactions in the Last Hour:")
+    account.getTransactionsBetween(LocalDateTime.now().minusHours(1), LocalDateTime.now()).forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Transactions:")
+    account.getAllSuccessfulTransactions().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Transactions:")
+    account.getAllFailedTransactions().forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Deposits:")
+    account.getAllSuccessfulDeposits().forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Withdrawals:")
+    account.getAllSuccessfulWithdrawals().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Deposits:")
+    account.getAllFailedDeposits().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Withdrawals:")
+    account.getAllFailedWithdrawals().forEach { println(it.getAmount()) }
+}
+
+class TransactionalBankAccount( accountNumber: String, accountHolderName: String, balance: Double = 0.0):
+    BankAccount(accountNumber, accountHolderName, balance) {
+
+    private var transactions = mutableListOf<Transaction>()
+
+    fun getAllTransactions(): List<Transaction> {
+        transactions.sortByDescending { it.getTransactionDate() }
+        return transactions
+    }
+    fun getAllTransactionsBy(predicate: (Transaction) -> Boolean) = transactions.filter(predicate)
+                                                                                .sortedByDescending{it.getTransactionDate()}
+    fun getTransactionsBetween(startDate: LocalDateTime, endDate: LocalDateTime) = getAllTransactionsBy { it.getTransactionDate() in startDate.rangeUntil(endDate) }
+
+    fun getAllFailedTransactions() = transactions.filter { it.getStatus() == TransactionStatus.FAILURE }
+    fun getAllSuccessfulTransactions() = transactions.filter { it.getStatus() == TransactionStatus.SUCCESS }
+
+    fun getAllFailedDeposits() = getAllFailedTransactions().filter { it.getTransactionType() == TransactionType.DEPOSIT }
+    fun getAllFailedWithdrawals() = getAllFailedTransactions().filter { it.getTransactionType() == TransactionType.WITHDRAW }
+    fun getAllSuccessfulDeposits() = getAllSuccessfulTransactions().filter { it.getTransactionType() == TransactionType.DEPOSIT }
+    fun getAllSuccessfulWithdrawals() = getAllSuccessfulTransactions().filter { it.getTransactionType() == TransactionType.WITHDRAW }
```

Comment:
None of these methods sorts transactions descending by date before returning the value.